cmd/main.go

```diff
@@ -18,64 +18,142 @@ package main
-
 	"k8s.io/klog/v2"
 )
 
+var (
+	featureGate = featuregate.NewFeatureGate()
+)
+
 func main() {
 	fs := flag.NewFlagSet("aws-ebs-csi-driver", flag.ExitOnError)
-
 	if err := logsapi.RegisterLogFormat(logsapi.JSONLogFormat, json.Factory{}, logsapi.LoggingBetaOptions); err != nil {
 		klog.ErrorS(err, "failed to register JSON log format")
 	}
 
-	options := GetOptions(fs)
+	var (
+		version  = fs.Bool("version", false, "Print the version and exit.")
+		toStderr = fs.Bool("logtostderr", false, "log to standard error instead of files. DEPRECATED: will be removed in a future release.")
+		args     = os.Args[1:]
+		cmd      = string(driver.AllMode)
+		options  = driver.Options{}
+	)
+
+	options.AddFlags(fs)
+
+	c := logsapi.NewLoggingConfiguration()
+	err := logsapi.AddFeatureGates(featureGate)
+	if err != nil {
+		klog.ErrorS(err, "failed to add feature gates")
+	}
+	logsapi.AddFlags(c, fs)
+
+	if len(os.Args) > 1 && !strings.HasPrefix(os.Args[1], "-") {
+		cmd = os.Args[1]
+		args = os.Args[2:]
+	}
+
+	switch cmd {
+	case "pre-stop-hook":
+		clientset, clientErr := metadata.DefaultKubernetesAPIClient()
+		if clientErr != nil {
+			klog.ErrorS(err, "unable to communicate with k8s API")
+		} else {
+			err = hooks.PreStop(clientset)
+			if err != nil {
+				klog.ErrorS(err, "failed to execute PreStop lifecycle hook")
+				klog.FlushAndExit(klog.ExitFlushTimeout, 1)
+			}
+		}
+		klog.FlushAndExit(klog.ExitFlushTimeout, 0)
+	case string(driver.ControllerMode), string(driver.NodeMode), string(driver.AllMode):
+		options.Mode = driver.Mode(cmd)
+	default:
+		klog.Errorf("Unknown driver mode %s: Expected %s, %s, %s, or pre-stop-hook", cmd, driver.ControllerMode, driver.NodeMode, driver.AllMode)
+		klog.FlushAndExit(klog.ExitFlushTimeout, 0)
+	}
+
+	if err = fs.Parse(args); err != nil {
+		panic(err)
+	}
+
+	err = logsapi.ValidateAndApply(c, featureGate)
+	if err != nil {
+		klog.ErrorS(err, "failed to validate and apply logging configuration")
+	}
+
+	if *version {
+		versionInfo, versionErr := driver.GetVersionJSON()
+		if versionErr != nil {
+			klog.ErrorS(err, "failed to get version")
+			klog.FlushAndExit(klog.ExitFlushTimeout, 1)
+		}
+		fmt.Println(versionInfo)
+		os.Exit(0)
+	}
+
+	if *toStderr {
+		klog.SetOutput(os.Stderr)
+	}
 
 	// Start tracing as soon as possible
-	if options.ServerOptions.EnableOtelTracing {
-		exporter, err := driver.InitOtelTracing()
-		if err != nil {
+	if options.EnableOtelTracing {
+		exporter, exporterErr := driver.InitOtelTracing()
+		if exporterErr != nil {
 			klog.ErrorS(err, "failed to initialize otel tracing")
 			klog.FlushAndExit(klog.ExitFlushTimeout, 1)
 		}
 		// Exporter will flush traces on shutdown
 		defer func() {
 			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 			defer cancel()
-			if err := exporter.Shutdown(ctx); err != nil {
+			if shutdownErr := exporter.Shutdown(ctx); shutdownErr != nil {
 				klog.ErrorS(err, "could not shutdown otel exporter")
```

Comment:
`err` -> `shutdownErr`

Reply:
```
cmd/main.go:124:7: shadow: declaration of "err" shadows declaration at line 59 (govet)
                        if err := exporter.Shutdown(ctx); err != nil {
```

Reply:
Edit requested change is:
```
klog.ErrorS(err, "could not shutdown otel exporter") -> klog.ErrorS(shutdownErr, "could not shutdown otel exporter")
```

making change next rev

Reply:
Done